### DISCUSSION TITLE
PV und Netzanschluss "Smart" machen mit zwei Shelly Pro 3 EM und Solectrus

### DISCUSSION THREAD

**Discussion by boecks** [solectrus/solectrus#3392] [Q&A] [answered]
Ich versuche gerade Solectrus neu aufzusetzen um PV Erzeugung und Haus Verbruach zu visualisieren.

Meine elektrische Sitauation ist wie folgt:
Ein Shelly Pro 3 em an den 3-Phasen direkt beim Hausanschluss als Haus Zähler für Energie Import und Export und
Ein Shelly Pro 3 em an der Steigleitung meiner PV Anlage (3-Phasen) um die PV erzeugung zu messen.
(eine besonderheit ist wohl, dass bei mir mehrere Wechselrichter an dieser Steigleitung hängen, daher dachte ich, an dem Messpunkt erhalte ich dann auch die Gesamte Produktion)

Kann Solecturs den Hauswerbruach aus den beiden gemessenen Werten automatisch berechnen oder muss das ein Drittsystem erledigen?

Ich verwende auf beiden Shelly ein Script welches die Aktive Leistung sekündlich via MQTT an Solecturs übermittelt. Aktuell finden die Daten der beiden Zähler erst in Solectrus zusammen, wenn ich vorab den Hausverbrauch errechnen müssten, dann wäre ich für Vorschläge dankbar. 

Das technische Umfeld bietet aktuell einen MQTT Broker, Solectrus mit MQTT Collector in Docker und die beiden Shelly Meter mit deren Möglichkeiten. 

In der Theorie könnte ich mit einem Shelly die Daten des anderen regelmäßig abfragen oder auf MQTT Topics reagieren und dann von diesem Shelly alle Solectrus Topics gesammelt übermitteln. Wenn euch was performanteres einfällt, dann sehr gerne!

LG Sascha

**Comment by ledermann** [comment]
Ich nehme an, es gibt bei Dir keinen Speicher, kein E-Auto und keine Wärmepumpe. Wenn ich das richtig verstanden habe, errechnet sich der Hausverbrauch dann wie folgt:

    Hausverbrauch = Netzbezug + Erzeugung - Einspeisung

Da die Werte von zwei unterschiedlichen Messstellen kommen (und somit nie gleichzeitig vorliegen), kann diese Berechnung derzeitig mit Bordmitteln nicht von vom Shelly-Collector oder dem MQTT-Collector durchgeführt werden. Letzterer kann [seit einiger Zeit zwar rechnen](https://github.com/solectrus/mqtt-collector/wiki/Konfiguration), aber nur innerhalb eines Topics, nicht topic-übergreifend.

Kurzfristig wäre das also nur machbar, wenn die Berechnung irgendwo außerhalb stattfindet, z.B. im ioBroker oder Home Assistant und dann über MQTT zu SOLECTRUS gelangt.

Es wäre aber denkbar, dass SOLECTRUS in einer zukünftigen Version diese Berechnung selbst durchführt. Eine Idee wäre, dass man `INFLUX_SENSOR_HOUSE_POWER` leer lässt und SOLECTRUS dann automatisch obige Formel anwendet. Wäre das korrekt? Und wäre das wünschenswert?

Noch allgemeiner: Grundsätzlich könnte man genau einen Wert (falls fehlend) aus den anderen errechnen - auf Basis der Strombilanz (Herkunft = Verwendung).

**Comment by boecks** [comment]
Du liegst richitg, es gibt keinen Speicher oder extra gemessene Verbraucher. Wobei Großverbraucher künfitg mit eigenen Zählern (wiederum Shelly) dazukommen könnten.

Für ein absolutes Basissetup wäre es für mich schon ein interessanter Ansatz, dass Solectrus ein default Verhalten für den Hausverbrauch hat, wenn nur Daten am Netzanschluss (Import/Export) und von der PV (Erzeugung) existieren. (es also auch keinen konfigurierten Sensor dafür gibt)

Leider existieren in diesem aktuellen Setup keine externen Smart Home Systeme wie HA etc. Wenn es nicht erforderlich ist, würde ich die auch gerne weglassen.

Der Ansatz den ich jetzt versuchen will ist, dass einer der beiden Shelly auserkoren wird, die Daten von sich und dem anderen Shelly zu sammeln, berechnungen zu machen und dann über die bestehenden MQTT Topics die Daten übermittelt. Ich denke von der Rechenleistung sollte das passen, vor allem, wenn ich zB das Intervall auf 5 Sekunden hochdrehe. Ich stelle mir vor, dass einer der Shelly auf ein Topic des anderen subscribed und via Script dann auf eine passende eingehende MQTT Nachricht reagiert, berechnungen macht und an Solectrus via MQTT published.

**Comment by boecks** [comment]
Wäre das nicht wiederum ein möglicher Use Case für den Power Splitter? Der dient aktuell ja schon dazu die Verbrauchsdaten live zu manipulieren.

**Comment by ledermann** [comment]
Für mich werde ich mal drei Erweiterungen notieren und separate Issues anlegen:

- Dashboard: Soll optional **einen** fehlenden Messwert aus den anderen berechnen können
- Shelly-Collector: Soll negative Werte optional als 0 schreiben
- Shelly-Collector: Soll optional Werte je nach Vorzeichen in unterschiedliche Fields schreiben können

Damit wäre dann dein Szenario abzubilden, ohne dass MQTT benötigt wird. Es wären dann zwei unterschiedlich konfigurierte Shelly-Collectoren im Einsatz, sonst nichts.

Der [PowerSplitter](https://github.com/solectrus/power-splitter) hat die Aufgabe, den Netzanteil von einem Stromverbrauch zu bestimmen. Du hast recht, das ist etwas Ähnliches. Aber das brauchen wir für dein Szenario vermutlich nicht bzw. wäre übertrieben aufwändig. Auch hat der PowerSplitter den Nachteil, dass er nur einmal pro Stunde rechnet. Das könnte man zwar hochdrehen, was aber den Traffic auf der InfluxDB sehr erhöhen würde, was ich gerne vermeiden möchte.

Die Differenzberechnung eines fehlenden Wertes (z.B. Hausverbrauch) dürfte recht einfach sein und könnte **ohne** expliziten Messwert in InfluxDB auskommen. Das wäre vergleichbar mit dem bereits vorhandenen Feature `INFLUX_EXCLUDE_FROM_HOUSE_POWER`. Naja, möglicherweise wird es bei der Top10-Hausverbrauch knifflig. Mal sehen. Notfalls müsste man auf die Top10 eines fehlenden Sensors verzichten, weil die InfluxDB-Performance für eine solche Abfrage zu schlecht wäre.

Bei deinem Lösungsansatz fällt mir noch auf, dass ein Shelly (soweit mir bekannt) nur MQTT-**Client** sein kann. Wenn du also in irgendeiner Weise mit MQTT hantierst, brauchst du zwingend irgendwo einen MQTT-Broker. Auch der MQTT-Collector verbindet sich nicht direkt mit einem Shelly, sondern nur mit einem MQTT-Broker. Wenn es kein Smarthome-System gibt, müsstest du auf etwas cloud-basiertes ausweichen, z.B. [HiveMQ](https://www.hivemq.com/).

**Comment by boecks** [comment]
Ich fände diese Erweiterungen sehr gut - das macht Energiemonitoring (inkl. Solectrus) kostentechnisch sehr niederschwellig zugänglich, da man bereits mit einem eher günsitgen Meter wie einem Shelly Pro 3 EM ein minimal Setup ohne großem Mehraufwand zustandebringt.

Bei mir läuft Mosquitto als MQTT Broker. Damit können beide Shelly und Solectrus über MQTT kommunizieren.

**Reply by ledermann** [reply]
Ok, dass Mosquitto lokal noch mit im Spiel ist, war mir nicht bewusst.

Deine Script-Lösung sorgt dafür, dass der Hausverbrauch als eigener Wert da ist, was für SOLECTRUS natürlich ideal ist, es muss an keiner Stelle irgendwas anders berechnet werden als sonst.

Meine oben skizzierten drei Erweiterungen sind für dich nun nicht mehr erforderlich. Ich werde mir dennoch weiter Gedanken dazu machen, weil man dadurch ohne MQTT auskommen würde.

**Comment by boecks** [comment]
Ich teste das gerade mittels Shelly Scripts wie beschrieben, kannst du meinen Gedanken bestätigen, dass bei kombiniertem Wert für Import und Export die Formel lautet: Hausverbrauch = Erzeugung - Netz wobei Netz bei Export einen positiven Wert hat und bei Import einen negativen?

Meine händischen Rechnungen würden das zumindest bestätigen.

**Reply by ledermann** [reply]
Ja, das sehe ich auch so, denn:

    Hausverbrauch = Netzbezug + Erzeugung - Einspeisung
    Netz          = Einspeisung - Netzbezug

    ⇒ Hausverbrauch = Erzeugung - Netz

**Comment by boecks** [comment, marked as answer]
Der gesamte Ansatz scheint zu funktionieren. Shelly 1 Publisht via MQTT, Shelly 2 reagiert auf diese Message indem er erstens die eigene gemessene Leistung abfragt, den Wert aus der Message von Shelly 1 entgegennimmt und daraus dann den Hausverbrauch berechnet. Shelly 2 publisht anschließend drei Messages auf den drei definierten MQTT Topics auf die auch Solectrus hört.

Für Interessierte die Scripte anbei. (bedarf möglicherweise individuelle Anpassungen)

Shelly A (liest die eigene gemessene Leistung aus und publisht den Wert in einem definiertem Zeitintervall - hier die Inverter Leistung)
```
// Set the MQTT topic and the publishing interval
const publishTopic = "shelly/meter/1/inverter/power";
const publishInterval = 5000;  // Publish every 5000 milliseconds (5 seconds)

// Function to retrieve the total active power data from this shelly
function getTotalActivePower() {
    return Shelly.getComponentStatus("em", 0).total_act_power;
}

// Timer to periodically publish the inverter power data
Timer.set(publishInterval, true, function() {
  MQTT.publish(publishTopic, JSON.stringify(getTotalActivePower()), 0, false);
});
```
Shelly B (nimmt den Wert von Shelly A entgegen, liest den eigenen Leistungswert aus und bereitet die Daten für Solectrus auf. Dieser Shelly könnte auch noch andere Messdaten abfragen, aufbereiten und publishen - zB ist hier noch eine Wärmepumpe vorbereitet)
```
// MQTT topics
const subscribeTopic = "shelly/meter/1/inverter/power";
const publishTopicInverter = "home/inverter/power";
const publishTopicHouse = "home/energy/house_power";
const publishTopicGrid = "home/grid/power";
const publishTopicHeatpump = "home/energy/heatpump1";

// Fetch grid power data
function getGridPower() {
    const gridPower = Shelly.getComponentStatus("em", 0).total_act_power;
    print("Grid Power:", gridPower); // Debug: Grid power
    return gridPower;
}

// Process inverter power data (ensure non-negative)
function processInverterPower(data) {
    const processedPower = data > 0 ? data : 0;
    print("Inverter Power (Raw/Processed):", data, "/", processedPower); // Debug: Inverter power
    return processedPower;
}

// Calculate house power
function calculateHousePower(inverterPower, gridPower) {
    const housePower = gridPower >= 0 ? inverterPower + gridPower : inverterPower - Math.abs(gridPower);
    print("House Power:", housePower); // Debug: House power
    return housePower;
}

// Process heatpump power data. currently dummy, since there is no meter installed
function getHeatpumpPower() {
    const queriedHeatpumpPower = 0;
    return queriedHeatpumpPower ;
}

// Handle incoming MQTT messages
function handleMqttMessage(topic, message) {
    print("Received MQTT Message for raw inverter power:", message); // Debug: Received message
    const inverterPower = processInverterPower(JSON.parse(message));
    const gridPower = getGridPower();
    const housePower = calculateHousePower(inverterPower, gridPower);
    const heatpumpPower = getHeatpumpPower();

    // Evaluate power state and publish data
    print(inverterPower > 0 ? "PV is producing power." : "");
    print(gridPower > 0 ? "Importing power from grid." : gridPower < 0 ? "Exporting PV power to grid." : "");
    MQTT.publish(publishTopicInverter, JSON.stringify(inverterPower), 0, false);
    MQTT.publish(publishTopicHouse, JSON.stringify(housePower), 0, false);
    MQTT.publish(publishTopicGrid, JSON.stringify(gridPower), 0, false);
    MQTT.publish(publishTopicHeatpump, JSON.stringify(heatpumpPower), 0, false);
}

// Subscribe to MQTT topic
MQTT.subscribe(subscribeTopic, handleMqttMessage);
```

EDIT 29.08.2024: Das beschriebene Setup läuft und funktinoiert einwandfrei mit dem MQTT Collector von Solectrus. Die beiden Scripte (wurden final aufgeräumt und inline dokumentiert) dienen als Anhaltspunkt für alle, die eine ähnliche Situation vorfinden und zB mit Shelly Meter auskommen müssen. Es wäre denkbar, dass weitere Shelly hinzukommen, welche dann auch von dem zentralen "Data Processor Shelly" vorbereitet und an Solectrus übermittelt werden. Der Vorteil an einem Setup wie diesem sind zum einen die geriengen Hardwarekosten und zum anderen, dass man sich Drittsoftware sparen kann, welche die Werte sonst aufbereiten würde. (Homeassistant, ioBroker, evcc, etc.) 
Solectrus und ein MQTT Broker reichen somit aus. Diesen Broker könnte man dann auch gleich mit Docker neben Solectrus deployen.

**Reply by ledermann** [reply]
Sehr, sehr coole Lösung! 💯
